### PR TITLE
refactor(Core/Packets): Rewrite various query packets to modern class.

### PR DESCRIPTION
**Pull request by heyitsbench**
## Changes Proposed:

This PR proposes changes to:
-  [X] Core (units, players, creatures, game systems).

## Issues Addressed:

- Closes unreported issue.

## Tests Performed:

This PR has been:
- [X]Tested in-game by the author.
- [X] This pull request requires further testing and may have edge cases to be tested.

## Known Issues and TODO List:

- [X] I honestly would like certain pieces to have some sort of boolean instead of things like `uint8(0)`, since makes more sense to a reader, but not sure if it'd have the same bit representation when sent as a packet, so also not sure if the client would accept it and work with it properly.

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

File: src/server/game/Server/WorldSession.h [hunk 1]
```diff
@@ -661,9 +668,9 @@ class WorldSession
     void HandleGameObjectUseOpcode(WorldPacket& recPacket);
     void HandleGameobjectReportUse(WorldPacket& recvPacket);
 
-    void HandleNameQueryOpcode(WorldPacket& recvPacket);
+    void HandleNameQueryOpcode(WorldPackets::Query::NameQuery& packet);
 
-    void HandleQueryTimeOpcode(WorldPacket& recvPacket);
+    void HandleQueryTimeOpcode(WorldPackets::Query::TimeQuery& packet);
```
Comment (sudlud) [inline src/server/game/Server/WorldSession.h]: Querytime? Instead of timequery, so it matches the function name and also the packet define

File: src/server/game/Handlers/QueryHandler.cpp [hunk 2]
```diff
@@ -32,50 +33,44 @@ void WorldSession::SendNameQueryOpcode(ObjectGuid guid)
...
-    data << uint8(0);                               // name known
-    data << playerData->Name;                       // played name
-    data << uint8(0);                               // realm name - only set for cross realm interaction (such as Battlegrounds)
-    data << uint8(player ? player->getRace() : playerData->Race);
-    data << uint8(playerData->Sex);
-    data << uint8(playerData->Class);
+    nameQueryResponse.NameUnknown = uint8(0);
+    nameQueryResponse.Name = playerData->Name;
+    nameQueryResponse.RealmName = uint8(0); // Only set for cross realm interaction (such as Battlegrounds)
+    nameQueryResponse.Race = uint8(player ? player->getRace() : playerData->Race);
```
Comment (sudlud) [inline src/server/game/Handlers/QueryHandler.cpp]: The castings when filling the member variables can be removed imo. The target data type is already known because of the member variable data type

File: src/server/game/Handlers/QueryHandler.cpp [hunk 3]
```diff
@@ -32,50 +33,44 @@ void WorldSession::SendNameQueryOpcode(ObjectGuid guid)
...
     Player* player = ObjectAccessor::FindConnectedPlayer(guid);
 
-    data << uint8(0);                               // name known
-    data << playerData->Name;                       // played name
-    data << uint8(0);                               // realm name - only set for cross realm interaction (such as Battlegrounds)
-    data << uint8(player ? player->getRace() : playerData->Race);
-    data << uint8(playerData->Sex);
-    data << uint8(playerData->Class);
+    nameQueryResponse.NameUnknown = uint8(0);
+    nameQueryResponse.Name = playerData->Name;
```
Comment (sudlud) [inline src/server/game/Handlers/QueryHandler.cpp:48]: This introduces an additional string copy, right?

File: src/server/game/Server/WorldSession.h (hunk 1 above)
Comment (heyitsbench) [reply src/server/game/Server/WorldSession.h]: I'd be more in favor of renaming the function to be in line with the NameQuery function above.

File: src/server/game/Handlers/QueryHandler.cpp (hunk 2 above)
Comment (heyitsbench) [reply src/server/game/Handlers/QueryHandler.cpp]: Should be corrected.

File: src/server/game/Handlers/QueryHandler.cpp (hunk 3 above)
Comment (heyitsbench) [reply src/server/game/Handlers/QueryHandler.cpp:48]: Believe so, is there an issue with that?

**Comment by heyitsbench** [comment]
Failing check seems to be due to [mod-war-effort](https://github.com/azerothcore/mod-war-effort), unrelated to this PR.

File: src/server/game/Handlers/QueryHandler.cpp (hunk 3 above)
Comment (Takenbacon) [reply src/server/game/Handlers/QueryHandler.cpp:48]: > Believe so, is there an issue with that?

It's always going to be preferable to avoid copies, you can probably pass it as a string view

Comment (heyitsbench) [reply src/server/game/Handlers/QueryHandler.cpp:48]: That's fair, I've replaced the datatype in the packet structure, retested and still seems to work fine for the client. I assume this is a suitable correction?

File: src/server/game/Server/WorldSession.h (hunk 1 above)
Comment (sudlud) [reply src/server/game/Server/WorldSession.h]: Yeah I think that's what I meant, having the same naming like the name query

File: src/server/game/Handlers/QueryHandler.cpp (hunk 3 above)
Comment (Takenbacon) [reply src/server/game/Handlers/QueryHandler.cpp:48]: > That's fair, I've replaced the datatype in the packet structure, retested and still seems to work fine for the client. I assume this is a suitable correction?

Yeah should be fine. The only stipulation with string views is they're basically a pointer to the string so there's lifetime considerations involved in the use, e.g. you'd have a problem using a string view on a temporary string that goes out of scope from where the string view is used. As long as the original string is valid in the same context it's safe and more efficient

File: src/server/game/Handlers/QueryHandler.cpp [hunk 4]
```diff
@@ -32,62 +33,55 @@ void WorldSession::SendNameQueryOpcode(ObjectGuid guid)
...
+        nameQueryResponse.NameUnknown = true;
+        SendPacket(nameQueryResponse.Write());
         return;
     }
 
     Player* player = ObjectAccessor::FindConnectedPlayer(guid);
 
-    data << uint8(0);                               // name known
-    data << playerData->Name;                       // played name
-    data << uint8(0);                               // realm name - only set for cross realm interaction (such as Battlegrounds)
```
Comment (heyitsbench) [inline src/server/game/Handlers/QueryHandler.cpp:48]: Down the road, this should be added back with something like this:

```cpp
if (player->HasSameNativeRealmAs(queryingPlayer))
    nameQueryResponse.RealmName = player->GetNativeRealmName();
```

But AC has no cross-realm support atm afaik, so these methods don't functionally exist. The client does properly show the (*) and the realm name on the player when this field is set, so nothing else needs doing it looks like to support that piece.

For now, I've just set the RealmName to be "" in the packet definition.

Comment (TheSCREWEDSoftware) [reply src/server/game/Handlers/QueryHandler.cpp:48]: Should this ^ (the above said) be added into the QueryHandler as TODO comment?

Comment (heyitsbench) [reply src/server/game/Handlers/QueryHandler.cpp:48]: I don't personally think so, I feel that the comment in QueryPackets.h conveys the nature of the field and what exactly it needs to be implemented.

File: src/server/game/Handlers/QueryHandler.cpp [hunk 5]
```diff
@@ -32,62 +33,55 @@ void WorldSession::SendNameQueryOpcode(ObjectGuid guid)
...
-    data << uint8(player ? player->getRace() : playerData->Race);
-    data << uint8(playerData->Sex);
-    data << uint8(playerData->Class);
+    nameQueryResponse.NameUnknown = false;
+    nameQueryResponse.Name = playerData->Name;
+    nameQueryResponse.Race = player ? player->getRace() : playerData->Race;
+    nameQueryResponse.Sex = player ? player->getGender() : playerData->Sex;
+    nameQueryResponse.Class = player ? player->getClass() : playerData->Class;
```
Comment (sudlud) [inline src/server/game/Handlers/QueryHandler.cpp:51]: This is a change to the previous implementation, on purpose?

Comment (heyitsbench) [reply src/server/game/Handlers/QueryHandler.cpp:51]: Yes, done for consistency across the attribute fields.
